src/libtcod/renderer_xterm.c

```diff
@@ -159,7 +174,130 @@ static void xterm_destructor(struct TCOD_Context* __restrict self) {
   xterm_cleanup();
 }
 
-TCOD_Context* TCOD_renderer_init_xterm(const char* window_title) {
+static int read_terminated_int(char term, int def) {
+  char buf[16] = "";
+  for (size_t i = 0; i < sizeof(buf) - 1; i++) {
+    int ch = getchar();
+    if (ch == term) {
+      buf[i] = '\0';
+      break;
+    }
+    if (!isdigit(ch)) {
+      return def;
+    }
+    buf[i] = ch;
+  }
+	return atoi(buf);
+}
+
+static void xterm_handle_input_escape() {
+  if (getchar() != '[')
+    return;
+  g_got_rows = read_terminated_int(';', 0);
+  g_got_columns = read_terminated_int('R', 0);
+  g_got_size_timestamp = SDL_GetTicks();
+}
+
+static int xterm_handle_input(void *arg) {
+  while (true) {
+    int ch = getchar();
+    if (ch == '\x1b') {
+      xterm_handle_input_escape();
+      continue;
+    }
+    int sym = ch;
+    int mod = KMOD_NONE;
+    if (isupper(ch)) {
+      sym = tolower(sym);
+      mod = KMOD_SHIFT;
+    }
+    SDL_Event down_event = {
+      .key = {
+        .type = SDL_KEYDOWN,
+        .timestamp = SDL_GetTicks(),
+        .windowID = 0,
+        .state = SDL_PRESSED,
+        .repeat = 0,
+        .keysym = {
+          .sym = sym,
+          .scancode = SDL_GetScancodeFromKey(sym),
+          .mod = mod
+        }
+      }
+    };
+    SDL_Event text_event = {
+      .text = {
+        .type = SDL_TEXTINPUT,
+        .timestamp = SDL_GetTicks(),
+        .windowID = 0,
+        .text[0] = ch,
+        .text[1] = '\0'
+      }
+    };
+    SDL_Event up_event = down_event;
+    up_event.type = SDL_KEYUP;
+    up_event.key.state = SDL_RELEASED;
+    SDL_PushEvent(&down_event);
+    SDL_PushEvent(&text_event);
+    SDL_PushEvent(&up_event);
+  }
+  return 0;
+}
+
+static TCOD_Error xterm_recommended_console_size(
+    struct TCOD_Context* __restrict self,
+    float magnification,
+    int* __restrict columns,
+    int* __restrict rows) {
+  fprintf(stdout, "\x1b[%i;%iH", INT_MAX, INT_MAX);
+  fflush(stdout);
+  Uint32 start_time = SDL_GetTicks();
+  fprintf(stdout, "\x1b[6n");
+  fflush(stdout);
+  while (!SDL_TICKS_PASSED(SDL_GetTicks(), start_time + 100)) {
+    if (SDL_TICKS_PASSED(g_got_size_timestamp, start_time) {
```

Comment:
I messed this up, it's hard to verify suggestions.
```suggestion
    if (SDL_TICKS_PASSED(g_got_size_timestamp, start_time)) {
```